source/declarations.tex

```diff
@@ -8355,74 +8081,47 @@
+    void mf();                  // the function \tcode{mf} and its type have \Cpp{} language linkage
+    void mf2(void(*)());        // the function \tcode{mf2} has \Cpp{} language linkage;
                                 // the parameter has type ``pointer to C function''
   };
 }
 \end{codeblock}
 \end{example}
 
 \pnum
-If two declarations declare functions with the same name and
-parameter-type-list\iref{dcl.fct} to be members of the same
-namespace or declare objects with the same name to be members of the same
-namespace and the declarations give the names different language linkages, the
-program is ill-formed; no diagnostic is required if the declarations appear in
-different translation units.
+If two declarations of an entity give it different language linkages, the
+program is ill-formed; no diagnostic is required if neither declaration
+is reachable from the other.
 \indextext{consistency!linkage specification}%
-Except for functions with \Cpp{} linkage, a function declaration without a
-linkage specification shall not precede the first linkage specification
-for that function. A function can be declared without a linkage
-specification after an explicit linkage specification has been seen; the
-linkage explicitly specified in the earlier declaration is not affected
-by such a function declaration.
+A redeclaration of an entity without a linkage specification
+inherits the language linkage of the entity and (if applicable) its type.
 
 \pnum
 \indextext{function!linkage specification overloaded}%
-At most one function with a particular name can have C language linkage.
-Two declarations for a function with C language linkage with the same
-function name (ignoring the namespace names that qualify it) that appear
-in different namespace scopes refer to the same function. Two
-declarations for a variable with C language linkage with the same name
-(ignoring the namespace names that qualify it) that appear in different
-namespace scopes refer to the same variable.
-An entity with C language linkage shall not be declared with the same name
-as a variable in global scope, unless both declarations denote the same entity;
-no diagnostic is required if the declarations appear in different translation units.
-A variable with C language linkage shall not be declared with the same name as a
-function with C language linkage (ignoring the namespace names that qualify the
-respective names); no diagnostic is required if the declarations appear in
-different translation units.
-\begin{note}
-Only
-one definition for an entity with a given name
-with C language linkage can appear in the
-program (see~\ref{basic.def.odr});
-this implies that such an entity
-must not be defined in more
-than one namespace scope.
-\end{note}
+Two declarations declare the same entity
+if they (re)introduce the same name,
+one declares a function or variable with C language linkage,
+and the other declares such an entity or a variable
+that belongs to the global scope-
```

Comment:
rewrite to clarify "the other" and "such an entity":
"Two declarations declare the same entity if they (re)introduce the same name, and one declares a function or variable with C language linkage and the other declares such an entity or declares a variable that belongs to the global scope."

i.e. "same name, one declares" -> "same name, and one declares"
and "linkage, and the other" -> "linkage and the other"

Reply:
added repetition of "declares" to disambiguate after talking to Davis

Reply:
It's better, but "such an entity" still isn't clear (because "such" still refers to the object of a separate clause; my rewrite addresses this by putting the 2nd and 3rd clauses together)

How about:
"Two declarations declare the same entity if they (re)introduce the same name, 
one declares a function or variable with C language linkage, and 
the other declares a function or variable with C language linkage or declares a variable that belongs to the global scope."
It's wordy but clear.

While we're here, might an index for "[declaring] the same entity" be useful?

Reply:
The only entity in view here is "a function or variable with C language linkage", so for me the referent of "such an entity" is clear. The wordy repetition seems unwarranted.

Reply:
Definitely deferring this for now; I'm not sure yet if this requires further action. Please elaborate on the deferral issue if you think so.